### PR TITLE
Port over non-redundant Span tests from corefxlab #28281

### PR DESCRIPTION
**Pull request by rossslaney**
Started work on issue #28281. I am a first-timer, so I may have missed entirely with my work. I am having trouble with cast tests from corefxlab. Thanks.

### CODE REVIEW

File: src/System.Memory/tests/Span/CovariantSliceNotSupported.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.MemoryTests;
+using Xunit;
+
+namespace System.SpanTests
+{
+    public static partial class SpanTests
+    {
+
```
Comment (danmosemsft) [inline src/System.Memory/tests/Span/CovariantSliceNotSupported.cs]: Nit please remove excess blank lines. Also in overflow.cs

**Comment by danmosemsft** [comment]
Thanks for the contribution @rossslaney

File: src/System.Memory/tests/Span/Overflow.cs [hunk 2]
```diff
@@ -60,6 +60,219 @@ public static void IndexOverflow()
...
+                unsafe
+                {
+                    int
+                        GuidThreeGiBLimit = (int)(ThreeGiB / sizeof(Guid)),
+                        GuidTwoGiBLimit = (int)(TwoGiB / sizeof(Guid)),
+                        GuidOneGiBLimit = (int)(OneGiB / sizeof(Guid));
+
+                    var span = new Span<Guid>((void*)huge, GuidThreeGiBLimit);
+                    Guid guid = Guid.NewGuid();
+                    var slice = span.Slice(GuidTwoGiBLimit + 1);
```
Comment (stephentoub) [inline src/System.Memory/tests/Span/Overflow.cs]: Nit: var => Span<Guid>

File: src/System.Memory/tests/Span/Overflow.cs [hunk 3]
```diff
@@ -60,6 +60,219 @@ public static void IndexOverflow()
...
+                    slice[0] = guid;
+                    slice = span.Slice(GuidOneGiBLimit).Slice(1).Slice(GuidOneGiBLimit);
+                    Assert.Equal(guid, slice[0]);
+                }
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(huge);
+            }
+        }
```
Comment (stephentoub) [inline src/System.Memory/tests/Span/Overflow.cs]: Same comments as previous test

File: src/System.Memory/tests/Span/Overflow.cs [hunk 4]
```diff
@@ -60,6 +60,219 @@ public static void IndexOverflow()
...
+
+        [Fact]
+        public static void ArrayCtorStartOverflow()
```
Comment (stephentoub) [inline src/System.Memory/tests/Span/Overflow.cs]: How does this name apply to this test? What does "overflow" here mean?

Also, what is this testing that doesn't already have tests?

File: src/System.Memory/tests/Span/Overflow.cs [hunk 5]
```diff
@@ -60,6 +60,219 @@ public static void IndexOverflow()
...
+
+                    slice = span.Slice(GuidOneGiBLimit).Slice(1).Slice(GuidOneGiBLimit);
+                    Assert.Equal(guid, slice[0]);
+                }
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(huge);
+            }
+        }
```
Comment (stephentoub) [inline src/System.Memory/tests/Span/Overflow.cs]: What's the difference with all of these "huge" tests? Why do we need them all?

File: src/System.Memory/tests/Span/Overflow.cs (hunk 4 above)
Comment (rossslaney) [reply src/System.Memory/tests/Span/Overflow.cs]: This may have been a redundant test.

File: src/System.Memory/tests/System.Memory.Tests.csproj [hunk 6]
```diff
@@ -20,6 +20,7 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="AllocationHelper.cs" />
+    <Compile Include="Span\CovariantSliceNotSupported.cs" />
```
Comment (ahsonkhan) [inline src/System.Memory/tests/System.Memory.Tests.csproj]: nit: move this to the other itemgroup with the rest of the span tests here:
https://github.com/dotnet/corefx/pull/30411/files#diff-00a03683c006a09943017a6311ec51c9R36

File: src/System.Memory/tests/Span/Overflow.cs (hunk 5 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/Span/Overflow.cs]: ReadOnlySliceStartUInt32Overflow and ReadOnlySliceStartLengthInt32Overflow are essentially identical. I don't think we need both.

File: src/System.Memory/tests/Span/Overflow.cs [hunk 7]
```diff
@@ -60,6 +60,208 @@ public static void IndexOverflow()
...
+
+        [Fact]
+        public static void ReadOnlyArrayCtorStartLengthOverflow()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Span/Overflow.cs]: I don't think this test belongs here. It isn't really an overflow, like the other tests.

It probably fits better with the AsSpan tests:
https://github.com/dotnet/corefx/blob/ec1671fd7feffe1070a94d9d86379e0dd95ddcdf/src/System.Memory/tests/Span/AsSpan.cs#L152

```C#
        [Fact]
        public void GuidArrayAsSpanWithStartAndLength()
        {
            var arr = new Guid[20];

            ReadOnlySpan<Guid> slice = arr.AsSpan().Slice(2, 2);
            Guid guid = Guid.NewGuid();
            slice[1] = guid;

            Assert.Equal(guid, arr[3]);
        }
```

File: src/System.Memory/tests/Span/CovariantSliceNotSupported.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+using System.MemoryTests;
+using Xunit;
+
+namespace System.SpanTests
+{
+    public static partial class SpanTests
+    {
+        [Fact]
+        public static void CovariantSlicesNotSupported1()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Span/CovariantSliceNotSupported.cs]: We already have such ctor tests for span (CovariantSlicesNotSupported1 / CovariantSlicesNotSupported3): https://github.com/dotnet/corefx/blob/ec1671fd7feffe1070a94d9d86379e0dd95ddcdf/src/System.Memory/tests/Span/CtorArray.cs#L97-L98

The only one that needs to be added is CovariantSlicesNotSupported2, which is testing AsSpan. Therefore, we should move that one with the rest of the AsSpan tests. 
CovariantAsSpanNotSupported


```C#
        [Fact]
        public static void CovariantAsSpanNotSupported()
        {
            object[] array = new string[10];
            Assert.Throws<ArrayTypeMismatchException>(() => array.AsSpan());
            Assert.Throws<ArrayTypeMismatchException>(() => array.AsSpan(0, a.Length));
        }
```

File: src/System.Memory/tests/System.Memory.Tests.csproj (hunk 6 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/System.Memory.Tests.csproj]: Actually remove this file all together.

**Comment by rossslaney** [comment]
@ahsonkhan I made the requested changes

File: src/System.Memory/tests/Span/Overflow.cs [hunk 9]
```diff
@@ -60,6 +60,157 @@ public static void IndexOverflow()
             }
         }
 
+        [Fact]
+        [OuterLoop()]
+        public static void SliceStartInt32Overflow()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Span/Overflow.cs]: Please use the same technique to allocate and use large memory as the other tests (i.e. use AllocationHelper, and skip on 32-bit platforms).

See examples here:
https://github.com/dotnet/corefx/blob/eb4bb7ff28d639dcd9fd0c505d78c6196a569f57/src/System.Memory/tests/Span/Clear.cs#L255-L300
https://github.com/rossslaney/corefx/blob/8b947b8f25498738e0b47314b2793827568865ec/src/System.Memory/tests/Span/Overflow.cs#L19-L61

Also, please make these specific to Windows/Mac only (as the exception may not get triggered on Linux).

File: src/System.Memory/tests/Span/Overflow.cs [hunk 10]
```diff
@@ -60,6 +60,157 @@ public static void IndexOverflow()
...
+                }
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(huge);
+            }
+        }
+
+        [Fact]
+        public static void SliceStartUInt32Overflow()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Span/Overflow.cs]: We do not need the UInt32 versions of these tests (SliceStartUInt32Overflow and ReadOnlySliceStartUInt32Overflow).

They are essentially identical to the Int32 versions and aren't testing different things. Therefore, removing them please.

The only difference is the following:
`ReadOnlySpan<Guid> slice = span.Slice((GuidTwoGiBLimit + 1));`
versus
`ReadOnlySpan<Guid> slice = span.Slice(GuidTwoGiBLimit + 1);`

File: src/System.Memory/tests/Span/Overflow.cs (hunk 9 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/Span/Overflow.cs]: Please feel free to ask questions if you have any or if you get stuck.

**Comment by rossslaney** [comment]
@ahsonkhan I made the requested changes. Please let me know if I missed anything.

**Comment by stephentoub** [comment]
Thanks.

**Pull request merged**